src/PrestaShopBundle/Controller/Admin/Improve/International/StateController.php

```diff
@@ -57,11 +69,177 @@ public function getStatesAction(Request $request)
+            'prestashop.core.form.identifiable_object.builder.state_form_builder'
+        );
+        $stateFormHandler = $this->get(
+            'prestashop.core.form.identifiable_object.handler.state_form_handler'
+        );
+
+        $stateForm = $stateFormBuilder->getForm();
+        $stateForm->handleRequest($request);
+
+        try {
+            $handlerResult = $stateFormHandler->handle($stateForm);
+            if ($handlerResult->isSubmitted() && $handlerResult->isValid()) {
+                $this->addFlash('success', $this->trans('Successful creation.', 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_states_index');
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+        }
+
+        return $this->render('@PrestaShop/Admin/Improve/International/Locations/State/add.html.twig', [
+            'enableSidebar' => true,
+            'stateForm' => $stateForm->createView(),
+            'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+        ]);
+    }
+
+    /**
+     * Handles edit form rendering and submission
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
+     *     redirectRoute="admin_states_index"
+     * )
+     *
+     * @param int $stateId
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function editAction(int $stateId, Request $request): Response
+    {
+        try {
+            /** @var EditableState $editableState */
+            $editableState = $this->getQueryBus()->handle(new GetStateForEditing((int) $stateId));
+        } catch (StateException $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+
+            return $this->redirectToRoute('admin_states_index');
+        }
+
+        $stateForm = null;
+
+        try {
+            $stateFormBuilder = $this->get(
+                'prestashop.core.form.identifiable_object.builder.state_form_builder'
+            );
+
+            $stateFormHandler = $this->get(
+                'prestashop.core.form.identifiable_object.handler.state_form_handler'
+            );
+
+            $stateForm = $stateFormBuilder->getFormFor((int) $stateId);
+            $stateForm->handleRequest($request);
+            $result = $stateFormHandler->handleFor((int) $stateId, $stateForm);
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_states_index');
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+
+            return $this->redirectToRoute('admin_states_index');
```

Comment:
Please only redirect for listing page if `$e` is an instance of `StateNotFoundException` 😉 
```
            if ($e instanceof StateNotFoundException) {
                return $this->redirectToRoute('admin_states_index');
            }
```

Reply:
refactored like this: 

```
        } catch (StateNotFoundException $e) {
            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));

            return $this->redirectToRoute('admin_states_index');
        } catch (Exception $e) {
            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
        }
```